tests/cuda/test_copy.cu

```diff
@@ -0,0 +1,153 @@
+traccc::device::container_d2h_copy_alg<test_container_types> d2h{mr, copy};
+
+}  // namespace
+
+TEST(CUDAContainerCopy, DeviceToHost) {
+
+    // Create cuda container
+    test_container_types::buffer device_buffer{{3, mr.main},
+                                               {{1, 3, 2}, mr.main, mr.host}};
+
+    // Get device vectors and fill them in the device
+    vecmem::device_vector<int> device_headers(device_buffer.headers);
+    vecmem::device_vector<int> device_items_0(
+        device_buffer.items.host_ptr()[0]);
+    vecmem::device_vector<int> device_items_1(
+        device_buffer.items.host_ptr()[1]);
+    vecmem::device_vector<int> device_items_2(
+        device_buffer.items.host_ptr()[2]);
+
+    thrust::fill(thrust::device, device_headers.begin(), device_headers.end(),
+                 7);
+    thrust::fill(thrust::device, device_items_0.begin(), device_items_0.end(),
+                 4);
+    thrust::fill(thrust::device, device_items_1.begin(), device_items_1.end(),
+                 1);
+    thrust::fill(thrust::device, device_items_2.begin(), device_items_2.end(),
+                 2);
+
+    // Device-to-Host Copy
+    test_container_types::host host_copy = d2h(device_buffer);
+
+    // Check the copied container
+    const auto& host_headers = host_copy.get_headers();
+    ASSERT_EQ(host_headers.size(), 3u);
+    ASSERT_EQ(host_headers[0], 7);
+    ASSERT_EQ(host_headers[1], 7);
+    ASSERT_EQ(host_headers[2], 7);
+
+    const auto& host_items = host_copy.get_items();
+    ASSERT_EQ(host_items.size(), 3u);
+    ASSERT_EQ(host_items[0].size(), 1u);
+    ASSERT_EQ(host_items[0][0], 4);
+    ASSERT_EQ(host_items[1].size(), 3u);
+    ASSERT_EQ(host_items[1][0], 1);
+    ASSERT_EQ(host_items[1][1], 1);
+    ASSERT_EQ(host_items[1][2], 1);
+    ASSERT_EQ(host_items[2].size(), 2u);
+    ASSERT_EQ(host_items[2][0], 2);
+    ASSERT_EQ(host_items[2][1], 2);
+}
+
+TEST(CUDAContainerCopy, DeviceToHostResizableBuffer) {
+
+    // Create cuda container
+    test_container_types::buffer device_buffer{
+        {3, mr.main},
+        {{1, 3, 2}, mr.main, mr.host, vecmem::data::buffer_type::resizable}};
+
+    // Get device vectors and fill them in the device
+    vecmem::device_vector<int> device_headers(device_buffer.headers);
+
+    thrust::fill(thrust::device, device_headers.begin(), device_headers.end(),
+                 7);
+
+    // Device-to-Host Copy
+    test_container_types::host host_copy = d2h(device_buffer);
+
+    // Check the copied container
+    const auto& host_headers = host_copy.get_headers();
+    ASSERT_EQ(host_headers.size(), 3u);
+    ASSERT_EQ(host_headers[0], 7);
+    ASSERT_EQ(host_headers[1], 7);
+    ASSERT_EQ(host_headers[2], 7);
+
+    const auto& host_items = host_copy.get_items();
+    ASSERT_EQ(host_items.size(), 3u);
+
+    // The test fails
+    ASSERT_EQ(host_items[0].size(), 0u);
+    ASSERT_EQ(host_items[1].size(), 0u);
+    ASSERT_EQ(host_items[2].size(), 0u);
```

Comment:
I think the problem you are having is because the d2h and h2d copy algorithms in traccc don't distinguish between resizable and fixed size vectors, so the final result object is being created with the size equal to the capacity of the input device buffer.

Reply:
whoops, I commented on the wrong PR

Reply:
Seems so. I am trying to fix that but it isn't easy